qiskit/quantum_info/analysis/z2_symmetries.py

```diff
@@ -0,0 +1,471 @@
+            )
+
+        stacked_matrix = np.stack(stacked_paulis)
+        symmetries = _kernel_F2(stacked_matrix)
+
+        if not symmetries:
+            logger.info("No symmetry is found.")
+            return cls([], [], [], None)
+
+        stacked_symmetries = np.stack(symmetries)
+        symm_shape = stacked_symmetries.shape
+        half_symm_shape = symm_shape[1] // 2
+        stacked_symm_del = [
+            np.delete(stacked_symmetries, row, axis=0) for row in range(symm_shape[0])
+        ]
+
+        def _test_symmetry_row_col(row: int, col: int, idx_test: list, row_test: list) -> bool:
+            """
+            Utilitary method that determines how to build the list of single-qubit Pauli X operators and
+            the list of corresponding qubit indices from the stacked symmetries.
+            This method is successively applied to Z type, X type and Y type symmetries (in this order)
+            to build the letter at position (col) of the Pauli word corresponding to the symmetry at
+            position (row).
+
+            Args:
+                row (int): Index of the symmetry for which the single-qubit Pauli X operator is being
+                    built.
+                col (int): Index of the letter in the Pauli word corresponding to the single-qubit Pauli
+                    X operator.
+                idx_test (list): List of possibilities for the stacked symmetries at all other rows
+                    than row and.
+                row_test (list): List of possibilities for the stacked symmetries at row.
+
+            Returns:
+                Whether or not this symmetry type should be used to build this letter of this
+                single-qubit Pauli X operator.
+            """
+            stacked_symm_idx_tests = np.array(
+                [
+                    (
+                        stacked_symm_del[row][symm_idx, col],
+                        stacked_symm_del[row][symm_idx, col + half_symm_shape],
+                    )
+                    in idx_test
+                    for symm_idx in range(symm_shape[0] - 1)
+                ]
+            )
+
+            stacked_symm_row_test = (
+                stacked_symmetries[row, col],
+                stacked_symmetries[row, col + half_symm_shape],
+            ) in row_test
+
+            return np.all(stacked_symm_idx_tests) and stacked_symm_row_test
+
+        for row in range(symm_shape[0]):
+            pauli_symmetries.append(
+                Pauli(
+                    (
+                        stacked_symmetries[row, :half_symm_shape],
+                        stacked_symmetries[row, half_symm_shape:],
+                    )
+                )
+            )
+
+            for col in range(half_symm_shape):
+                # case symmetries other than one at (row) have Z or I on col qubit
+                Z_or_I_idx_test = [(0, 0), (0, 1)]
+                Z_or_I_row_test = [(1, 0), (1, 1)]
```

Comment:
Maybe we can avoid re-defining these inside the inner-most loop and instead define them once outside the outer-most loop?
Same for the `X` and `Y` cases